### ISSUE TITLE
gcc build issue with CHECK_* macros

### ISSUE THREAD

**Issue by danluu** [BitFunnel/BitFunnel#225]
I get a 4 page long screen of error messages which contains:

```
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [11]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[11]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [11]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:197:16:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
         if (value == nullptr)
             ~~~~~~^~~~~~~~~~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [13]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[13]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [13]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:199:16:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [16]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[16]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [16]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:134:24:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [5]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_conv
ertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[5]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [5]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:137:24:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [15]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[15]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [15]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:137:41:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [7]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_
...
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [2]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_conv
ertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[2]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [2]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:160:35:   required from 'Logging::CheckException Logging::CheckResult<T1, T2>::Exception(const char*, const char*, const char*) 
const [with T1 = BitFunnel::ByteCodeInterpreter::Opcode; T2 = BitFunnel::ByteCodeInterpreter::Opcode]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [13]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [16]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [5]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [15]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [7]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
...
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [2]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:0:
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h: In constructor 'BitFunnel::ByteCodeInterpreter::Instruction::Instruction(BitFunnel::ByteCodeInterpreter::Opcode, size_t, 
size_t, bool)':
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:220:43: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
         out << static_cast<unsigned>(value);
                                           ^
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:252:45: note: '<anonymous>' was declared here
     ::Logging::CmpHelper##op_name(val1, val2))       \
                                             ^
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:265:5: note: in expansion of macro 'CHECK_OP'
     CHECK_OP(val1, val2, LT)
     ^~~~~~~~
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17: note: in expansion of macro 'CHECK_LT'
                 CHECK_LT(opcode, Opcode::Last)
                 ^~~~~~~~
```

I'll probably take a look at this later tonight or sometime tomorrow.

**Comment by danluu** [comment]
After 0fa6fe6, we're down to this, much shorter, error:

~~~
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:0:
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h: In constructor 'BitFunnel::ByteCodeInterpreter::Instruction::Instruction(BitFunnel::ByteCodeInterpreter::Opcode, size_t, size_t, bool)':
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:220:43: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
         out << static_cast<unsigned>(value);
                                           ^
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:246:45: note: '<anonymous>' was declared here
     ::Logging::CmpHelper##op_name(val1, val2))       \
                                             ^
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:259:5: note: in expansion of macro 'CHECK_OP'
     CHECK_OP(val1, val2, LT)
     ^~~~~~~~
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17: note: in expansion of macro 'CHECK_LT'
                 CHECK_LT(opcode, Opcode::Last)
                 ^~~~~~~~
~~~

**Comment by hausdorff** [comment]
Continuing the conversation from #258, the above is caused by a [GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=40146). One solution is to have GCC `#pragma`s around the affected lines, which disable `-Wmaybe-initialized`.

**Comment by MikeHopcroft** [comment]
Dan was aware of the GCC bug, but didn't have confirmation that the bug was actually responsible for the problem (one issue is that the problem occurs on versions of GCC where the bug should be resolved).

We don't really know the risks due to suppressing the error.

We do know that gtest does not have this problem and we are doing the same sort of thing they are doing. It would be interesting to look at their equivalent code and find out what work-around they used. Is this something you could investigate?

**Comment by hausdorff** [comment]
At commit [`0fa6fe6`](https://github.com/BitFunnel/BitFunnel/commit/0fa6fe662e9798330a232693ae7621160672ea2a), the error is caused by [line 218 of `ByteCodeInterpreter.h`](https://github.com/BitFunnel/BitFunnel/blob/0fa6fe662e9798330a232693ae7621160672ea2a/src/Plan/src/ByteCodeInterpreter.h#L218), which defines a function:

```c++
inline std::ostream& operator<<(std::ostream& out, const ByteCodeInterpreter::Opcode value)
{
    out << static_cast<unsigned>(value);
    return out;
}
```

Changing that signature to this:

```c++
inline std::ostream& operator<<(std::ostream& out, const ByteCodeInterpreter::Opcode& value)
```

causes the error to disappear.

To me, this looks like a compiler bug, but C++ is complicated, and it is possible that I am missing something subtle about the semantics of formal parameters that are references. What do you think?

If you think this looks like a bug, then my recommended next steps are to look through the GCC bug tracker, see whether this is known to be regressed, and if it's not, report it.

**Comment by hausdorff** [comment]
Ok, chatting with some GCC folks and the BitFunnel team out-of-band, I think we have enough information that we can probably safely suppress the warning Dan encountered above. My recommendation is that I create a patch and we close the issue. The reasons, in brief:

* The warning does not appear on the SVN trunk of GCC, so filing a bug against GCC would cause it to be marked as `FIXED WORKSFORME`.
* I believe the bug was fixed with [66977](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66977). After some playing around with the warnings, I learned that the problem is caused by our use of the `<<` operator in connection with the `CHECK_OP` macro; this bug specifically fixes a known problem with `-fsanitize=shift` that causes phantom uninitialized variables to appear. Further, I've confirmed the warning does show up on g++ version 5.4.0, but disappears on 5.4.1 and later, which seems to corroborate this theory.
* In addition to the above, I have thoroughly inspected the code and can detect no uninitialized variables.

Context and background
-----

Building on a more recent head ([`d5a9e`](https://github.com/BitFunnel/BitFunnel/commit/d5a9e649dbcae419e8395da6e43c37f711b9594e)) and suppressing Dan's warning from above reveals that this is actually a _family_ of 5 or so warnings, all of which are caused by two specific variable references.

A bit of context. Each of all of these warnings terminates the same way: the [`CHECK_OP`](https://github.com/BitFunnel/BitFunnel/blob/d5a9e649dbcae419e8395da6e43c37f711b9594e/inc/LoggerInterfaces/Check.h#L244) macro is expanded, and then, somehow, the actual warning is generated in a seemingly unrelated place in the source code.

For example, in Dan's error above, the warning message says that `CHECK_OP` is the last macro to be expanded, but then the actual warning is generated by [line 257 of `ByteCodeInterpreter.h`](https://github.com/BitFunnel/BitFunnel/blob/d5a9e649dbcae419e8395da6e43c37f711b9594e/src/Plan/src/ByteCodeInterpreter.h#L257). At first, his doesn't seem to be connected to `CHECK_OP` at all. But also, the warning complains that a symbol, `value`, might be uninitialized, but (as I mentioned earlier) this does not seem to be possible, looking at the code:

```c++
inline std::ostream& operator<<(std::ostream& out, const ByteCodeInterpreter::Opcode value)
{
    out << c_opcodeNames[static_cast<size_t>(value)];
    return out;
}
```

If you suppress the warning around that line, you end up with a family of errors complaining that the second argument of `StreamOut` ([line 45 of `Check.h`](https://github.com/BitFunnel/BitFunnel/blob/d5a9e649dbcae419e8395da6e43c37f711b9594e/inc/LoggerInterfaces/Check.h#L45)) might not be initialized.


The problem
-----

The root of the problem is our use of the `<<` operator in connection with the `CHECK_OP` macro.

1. `CHECK_OP` is designed so that you can use the `<<` operator to stream a message in case of error. For example you might right `CHECK_OP(opcode, Opcode:AndRow, GT) << "uh oh, the check failed!"`, and that message would get recorded if the check fails.
2. The problem in all of these cases comes from this use of `<<`. When `CHECK_OP` expands out, somewhere inside it calls `<<` in some ways that bother the compiler. Here's where it gets interesting.
3. Dan's error is caused in the method `inline std::ostream& operator<<(std::ostream& out, const ByteCodeInterpreter::Opcode value)`; in other words, it fails when we shift an `Opcode` into an `ostream`. In the errors we observed later, `StreamOut`, something similar holds, but with different operand types.

In order to confirm that this is the root of the error, the task at hand is now to determine: **where are these bad shifts happening**?

Well, it turns out that under the covers `CHECK_OP` expands to a call to `CheckResult.Exception` ([line 147 of `Check.h`](https://github.com/BitFunnel/BitFunnel/blob/d5a9e649dbcae419e8395da6e43c37f711b9594e/inc/LoggerInterfaces/Check.h#L147)); it is in this function that all our problems lie. To resolve all errors, simply change this function from this:

```c++
CheckException Exception(char const* expr1,
                         char const* expr2,
                         char const* op) const
{
    return CheckException()
        << "Expected: (" << expr1 << ") " << op
        << " (" << expr2 << "), actual: (" << m_val1
        << ") vs (" << m_val2 << ")" << std::endl;
}
```

to this:

```c++
CheckException Exception(char const* expr1,
                         char const* expr2,
                         char const* op) const
{
    return CheckException()
        << "Expected: (" << expr1 << ") " << op
        << " (" << expr2 << "), actual: (" /*<< m_val1*/
        << ") vs (" /*<< m_val2*/ << ")" << std::endl;
}
```

And that completely eliminates the entire family of errors. You might think: ok, so we're not initializing those variables? Nope, it turns out we are. They're part of the initialization list ([line 142 of `Check.h`](https://github.com/BitFunnel/BitFunnel/blob/d5a9e649dbcae419e8395da6e43c37f711b9594e/inc/LoggerInterfaces/Check.h#L142)). So this should not have caused an error.


Recommendations.
----
As I said above, my recommendation is pretty simple. I think we should suppress these warnings. I'm personally pretty convinced this is a GCC bug and I think we incur no risk by doing this.

**Comment by hausdorff** [comment]
Following up, we've also tried marking the default constructor to `CheckResult` as private, and implementing the copy constructor explicitly, so that we're sure the problem isn't that the member variables `m_val1` and `m_val2` aren't defined somehow.

At this point I'm about 95% sure the right thing to do is to suppress the warnings on the relevant lines. And, I'm about 98% sure that this is a compiler bug. Hence, I have submitted the PR.
